### ISSUE TITLE
[analyzer] ToSourceVisitor visitTopLevelVariableDeclaration misses visiting metadata 

### ISSUE THREAD

**Issue by incendial** [dart-lang/sdk#51236]
Hi, I've noticed that `ToSourceVisitor` [does not visit metadata of top level variables declarations](https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/lib/src/dart/ast/to_source_visitor.dart#L1316). Considering, that the metadata is visited for other nodes, this looks like a bug. 

Code example:
```dart
class SomeAnnotation {
  const SomeAnnotation();
}

const ignoredAnnotation = SomeAnnotation();

@ignoredAnnotation
void globalFunction() {}

@ignoredAnnotation
var globalVariable = 42;

@ignoredAnnotation
final globalFinalVariable = 42;
```

Produces:

<img width="776" alt="Screenshot 2023-02-02 at 23 28 04" src="https://user-images.githubusercontent.com/21116742/216430539-ad8fc5d1-e0fc-400c-9736-37d5cbbb68f0.png">

If the issue is confirmed, I can create a PR and a test for it.

**Comment by bwilkerson** [comment]
It's _sort_ of a bug, but a super low priority bug. The `ToSourceVisitor` is only used by `AstNode.toSource`, but that method only exists for debugging purposes, so nothing should depend on it being correct.

That said, a fix would be nice.

**Comment by incendial** [comment]
Hm, but calling `toString()` on any node also calls `toSource()` under the hood. Or we should not call `toString` on the ast nodes?

> That said, a fix would be nice.

👍

**Comment by bwilkerson** [comment]
We reserve `toString()` for debugging purposes everywhere in the analyzer APIs.

**Comment by incendial** [comment]
Another question then (sorry 😁): maybe you have a good approach to compare two nodes without `toString`? For example, two declarations.

**Comment by bwilkerson** [comment]
I'm not aware of any comparison support that we have. I can't think of any use case for it. What problem are you trying to solve?

**Comment by incendial** [comment]
Since you mentioned that `toString` is actually for debug purposes, we have several usages of it in our rules (for example, comparing that `then` and `else` statements are equal https://github.com/dart-code-checker/dart-code-metrics/blob/master/lib/src/analyzers/lint_analyzer/rules/rules_list/no_equal_then_else/visitor.dart#L23). And it's used in several places for the same result (to understand if two nodes have the same ast).

We also use `toString` in quick fixes, but I'm completely reworking the whole approach (I'll create another issue soon, looks like there are some ideas that can be reused), so that's not a problem.

**Comment by bwilkerson** [comment]
I can see how it would be useful for that. While we don't make any guarantees that it will always be appropriate for that, it seems unlikely that we'd change it in a way that would break you.

I definitely wouldn't use `toString` in a fix. The approach we usually take is to copy a range of text (based on the AST `offset` and `length` info) from the original source code. That preserves comments, formatting, etc.

**Comment by incendial** [comment]
Done: https://dart-review.googlesource.com/c/sdk/+/280468

> I definitely wouldn't use toString in a fix. The approach we usually take is to copy a range of text (based on the AST offset and length info) from the original source code. That preserves comments, formatting, etc.

I actually would like to discuss that, but in a separate issue

**Comment by asashour** [comment]
I think this is fixed.

To have a commit automatically closes a relevant issue, the message can be [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to it, for example `fixes #1234`
